File: examples/workflow-graph-roundtrip/frontend/src/lib/operations.js
// The operation catalog is host-owned data: the backend's `GET /operations` is
// the single source of truth for what a palette can insert and how each node's
// typed fields are shaped (ADR 0026 stance — no client-side catalog to drift).
// If the endpoint is unreachable the palette shows an error/empty state rather
// than falling back to a duplicated hard-coded list.
//
// A catalog entry is
//   { id, label, nodeKind, subkind?, operation?, effect?, terminalKind?,
//     fields: [{ name, type, default }] }
// where `type` is string | number | boolean | expression | identifier |
// assignment_target.

import { NODE_KINDS, CONTAINER_SUBKINDS, kindMeta } from './nodeKinds.js';

// Palette groups, in menu order. Each entry lands in the first group whose
// `kinds` include its nodeKind; unmatched kinds fall into "Other".
const GROUPS = [
  { id: 'actions', label: 'Actions', kinds: ['call', 'effect'] },
  { id: 'control', label: 'Control flow', kinds: ['container', 'terminal'] },
  { id: 'data', label: 'Data', kinds: ['data', 'computation', 'state_update'] },
  { id: 'process', label: 'Process', kinds: ['process'] },
  { id: 'advanced', label: 'Advanced', kinds: ['opaque'] },
];

// Kinds hidden from the Simplified palette (raw escape hatches).
const POWER_KINDS = new Set(['opaque']);

// Kinds that may only be inserted at the top level, not into a container slot.
const TOP_LEVEL_KINDS = new Set(['process']);

// Group a catalog into `[{ id, label, items }]`, dropping empty groups.
// Returns `[]` for a missing/empty catalog so the palette can show an empty
// state. Simplified mode hides raw power-only kinds; `topLevel:false` (a
// container's add-menu) additionally hides top-level-only kinds like `process`.
export function groupOperations(catalog, { includePower = true, topLevel = true } = {}) {
  const entries = (catalog ?? []).filter(
    (op) =>
      (includePower || !POWER_KINDS.has(op.nodeKind)) && (topLevel || !TOP_LEVEL_KINDS.has(op.nodeKind)),
  );
  if (!entries.length) return [];
  const groups = GROUPS.map((g) => ({
    id: g.id,
    label: g.label,
    items: entries.filter((op) => g.kinds.includes(op.nodeKind)),
  }));
  const claimed = new Set(GROUPS.flatMap((g) => g.kinds));
  const other = entries.filter((op) => !claimed.has(op.nodeKind));
  if (other.length) groups.push({ id: 'other', label: 'Other', items: other });
  return groups.filter((g) => g.items.length);
}

// The catalog entries a node of `nodeKind` can switch between (its operation
// options) — display calls for `call`, sleep/wait_signal for `effect`, etc.
export function operationsForKind(catalog, nodeKind) {
  return (catalog ?? []).filter((op) => op.nodeKind === nodeKind);
}

// A single field's default as an EditableValue for the `data.fields` map.
export function fieldDefaultValue(field) {
  switch (field.type) {
    case 'number':
      return Number(field.default ?? 0) || 0;
    case 'boolean':
      return !!field.default;
    case 'expression':
      return { $expr: String(field.default ?? '') };
    default:
      return String(field.default ?? '');
  }
}

// The seed `data.fields` map for an operation entry (used to prefill the arg
// form when a node is inserted or switched to a different operation).
export function catalogFieldsMap(op) {
  const out = {};
  for (const field of op.fields ?? []) out[field.name] = fieldDefaultValue(field);
  return out;
}

// The `data` patch to apply when switching an existing call/effect to a
// different operation: point at the new operation (calls swap the receiver via
// `operation`; effects rebuild from `effect` + fields, so drop any stored
// `expression`) and refill the arg form from the new operation's typed defaults.
export function operationSwitchPatch(nodeKind, op) {
  const patch = { fields: catalogFieldsMap(op) };
  if (nodeKind === 'call') patch.operation = op.operation;
  else if (nodeKind === 'effect') {
    patch.effect = op.effect;
    patch.clearExpression = true;
  }
  return patch;
}

// The catalog entry a node currently matches, so an operation `<select>` can
// preselect it: calls key on `operation`, effects on `effect`, terminals on
// `terminalKind`.
export function currentOperationId(catalog, node) {
  const kind = node?.data?.kind;
  const match = (predicate) => (catalog ?? []).find(predicate)?.id ?? null;
  if (kind === 'call') return match((op) => op.operation === node.data.operation);
  if (kind === 'effect') return match((op) => op.effect === node.data.effect);
  if (kind === 'terminal') return match((op) => op.terminalKind === node.data.terminalKind);
  return null;
}

// Menu glyph + accent for a catalog entry (containers borrow the container
// accent + their sub-kind glyph, so the palette reads like the canvas legend).
export function operationMeta(op) {
  if (op.nodeKind === 'container' && op.subkind) {
    const sub = CONTAINER_SUBKINDS[op.subkind] ?? CONTAINER_SUBKINDS.loop;
    return { glyph: sub.glyph, accent: NODE_KINDS.container.accent };
  }
  const meta = NODE_KINDS[op.nodeKind] ?? kindMeta(op.nodeKind);
  return { glyph: meta.glyph, accent: meta.accent };
}

File: examples/workflow-graph-roundtrip/src/operations.rs
use serde_json::{Value, json};

use crate::display;
use crate::{OperationCatalogEntry, OperationField};

pub(crate) fn entries() -> Vec<OperationCatalogEntry> {
    let mut entries = display::OPERATIONS
        .iter()
        .map(|operation| OperationCatalogEntry {
            id: format!("display.{}", operation.operation),
            label: operation.label.to_string(),
            node_kind: "call".to_string(),
            subkind: None,
            operation: Some(operation.operation.to_string()),
            effect: None,
            terminal_kind: None,
            fields: operation
                .fields
                .iter()
                .map(|field| OperationField {
                    name: field.name.to_string(),
                    field_type: field.field_type.to_string(),
                    default: display_default(field.field_type),
                })
                .collect(),
        })
        .collect::<Vec<_>>();
    entries.extend([
        entry(
            "effect.sleep",
            "Sleep",
            "effect",
            None,
            None,
            Some("sleep"),
            None,
            vec![field("duration", "expression", json!("\"1s\""))],
        ),
        entry(
            "effect.wait_signal",
            "Wait for signal",
            "effect",
            None,
            None,
            Some("wait_signal"),
            None,
            vec![field("signal", "string", json!("continue"))],
        ),
        entry(
            "control.if",
            "If / branch",
            "container",
            Some("if"),
            None,
            None,
            None,
            vec![field("condition", "expression", json!("true"))],
        ),
        entry(
            "control.while",
            "While loop",
            "container",
            Some("while"),
            None,
            None,
            None,
            vec![field("condition", "expression", json!("false"))],
        ),
        entry(
            "control.for",
            "For each",
            "container",
            Some("for"),
            None,
            None,
            None,
            vec![
                field("binding", "identifier", json!("item")),
                field("iterable", "expression", json!("[1, 2, 3]")),
            ],
        ),
        entry(
            "stmt.assign",
            "Set variable",
            "state_update",
            None,
            None,
            None,
            None,
            vec![
                field("target", "assignment_target", json!("state.count")),
                field("expression", "expression", json!("0")),
            ],
        ),
        entry(
            "stmt.let",
            "Define value",
            "data",
            None,
            None,
            None,
            None,
            vec![
                field("binding", "identifier", json!("value")),
                field("expression", "expression", json!("0")),
            ],
        ),
        entry(
            "stmt.compute",
            "Compute",
            "computation",
            None,
            None,
            None,
            None,
            vec![field("expression", "expression", json!("1 + 1"))],
        ),
        entry(
            "stmt.finish",
            "Finish",
            "terminal",
            None,
            None,
            None,
            Some("finish"),
            vec![field("expression", "expression", json!("0"))],
        ),
    ]);
    entries
}

fn display_default(field_type: &str) -> Value {
    match field_type {
        "number" => json!(0),
        _ => json!(""),
    }
}

#[allow(clippy::too_many_arguments)]
fn entry(
    id: &str,
    label: &str,
    node_kind: &str,
    subkind: Option<&str>,
    operation: Option<&str>,
    effect: Option<&str>,
    terminal_kind: Option<&str>,
    fields: Vec<OperationField>,
) -> OperationCatalogEntry {
    OperationCatalogEntry {
        id: id.to_string(),
        label: label.to_string(),
        node_kind: node_kind.to_string(),
        subkind: subkind.map(str::to_string),
        operation: operation.map(str::to_string),
        effect: effect.map(str::to_string),
        terminal_kind: terminal_kind.map(str::to_string),
        fields,
    }
}

fn field(name: &str, field_type: &str, default: Value) -> OperationField {
    OperationField {
        name: name.to_string(),
        field_type: field_type.to_string(),
        default,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_display_tool_has_exactly_one_catalog_entry() {
        let entries = entries();
        let display_entries = entries
            .iter()
            .filter(|entry| entry.id.starts_with("display."))
            .collect::<Vec<_>>();
        assert_eq!(display_entries.len(), display::OPERATIONS.len());
        for operation in display::OPERATIONS {
            let entry = display_entries
                .iter()
                .find(|entry| entry.operation.as_deref() == Some(operation.operation))
                .expect("display operation catalog entry");
            assert_eq!(entry.id, format!("display.{}", operation.operation));
            assert_eq!(entry.fields.len(), operation.fields.len());
        }
    }
}
